crates/core/src/transport/udp_connection.rs

```diff
@@ -0,0 +1,187 @@
+                    Some((addr, message)) = decrypted_packets_receiver_clone.recv() => {
+                        // Handle the decrypted packet
+                        Self::handle_decrypted_packet(&addr, &message).await;
+                    },
+                }
+            }
+        });
+
+        Ok(connection)
+    }
+
+    async fn handle_raw_packet(addr: &SocketAddr, message: &RawPacket) {
+        match message {
+            RawPacket::Message(data) => {
+                // Decrypt the message
+                let decrypted_message = self.decrypt_message(data).unwrap();
+
+                // Send the decrypted message to the decrypted packets channel
+                if let Err(e) = self
+                    .decrypted_packets
+                    .0
+                    .send((addr.clone(), decrypted_message))
+                    .await
+                {
+                    tracing::warn!("Failed to send decrypted message: {:?}", e);
+                }
+            }
+            RawPacket::Terminate => {
+                todo!()
+            }
+        }
+    }
+
+    async fn handle_decrypted_packet(addr: &SocketAddr, message: &Vec<u8>) {
+        todo!()
+    }
+
+    // Encrypts the data and prepends the nonce to the ciphertext
+    fn encrypt_message(&self, data: &[u8]) -> Result<Vec<u8>, ConnectionError> {
+        let mut nonce = [0u8; 12]; // 12 bytes nonce for AES-GCM
+        OsRng.fill_bytes(&mut nonce);
+
+        let cipher = self
+            .outbound_symmetric_key
+            .ok_or(ConnectionError::ProtocolError(
+                "Don't have outbound symmetric key".to_string(),
+            ))?;
+        let encrypted_data = cipher
+            .encrypt(GenericArray::from_slice(&nonce), data)
+            .map_err(|e| ConnectionError::AesGcmError(e))?;
+
+        // Prepend the nonce to the ciphertext
+        let mut result = Vec::with_capacity(nonce.len() + encrypted_data.len());
+        result.extend_from_slice(&nonce);
+        result.extend_from_slice(&encrypted_data);
+
+        Ok(result)
+    }
+
+    // Decrypts the data, assuming the nonce is prepended to the ciphertext
+    fn decrypt_message(&self, data: &[u8]) -> Result<Vec<u8>, ConnectionError> {
+        // Extract the nonce from the beginning of the data
+        if data.len() < 12 {
+            return Err(ConnectionError::ProtocolError(
+                "Data is too short to contain nonce".to_string(),
+            ));
+        }
+        let nonce = &data[..12];
+        let ciphertext = &data[12..];
+
+        let cipher = self
+            .inbound_symmetric_key
+            .ok_or(ConnectionError::ProtocolError(
+                "Don't have inbound symmetric key".to_string(),
+            ))?;
+        cipher
+            .decrypt(GenericArray::from_slice(nonce), ciphertext)
+            .map_err(|e| ConnectionError::AesGcmError(e))
+    }
+}
+
+enum RawPacket {
+    Message(Vec<u8>),
+    Terminate,
+}
+
+pub struct PacketQueue<T> {
```

Comment:
this is also a smell
sender shouldn't be needed to be locked and shared like this since you just can clone the sender side
and if receiver has to be shared is a sign of misuse since it should be consumed in one single place (otherwise what's the point, and you may be running into racey behaviour if you are sharing the receiver side)

But again I think all the multithreaded code here must go away.

Reply:
Yeah, this is old code from before I realized I had the wrong approach.